CLOUD-526 Document the Operator API (#518)
percona/percona-xtradb-cluster-operator: source/api.rst

   
   .toggle .header.open:after {
       content: " ▼";
   }
   </style>

Percona Operator for Percona XtraDB Cluster provides an `aggregation-layer extension for the Kubernetes API <https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/>`_. Please refer to the
`official Kubernetes API documentation <https://kubernetes.io/docs/reference/>`_ on the API access and usage details.
The following subsections describe the Percona XtraDB Cluster API provided by the Operator.

.. contents:: :local:

Prerequisites
-------------

1. prepare

   .. code-block:: yaml

      # set correct API adress
      KUBE_CLUSTER=$(kubectl config view --minify -o jsonpath='{.clusters[0].name}')
      API_SERVER=$(kubectl config view -o jsonpath="{.clusters[?(@.name==\"$KUBE_CLUSTER\")].cluster.server}")

      # create service account and get token
      kubectl apply -f deploy/crd.yaml -f deploy/rbac.yaml
      KUBE_TOKEN=$(kubectl get secret $(kubectl get serviceaccount percona-xtradb-cluster-operator -o jsonpath='{.secrets[0].name}' -n $NAMESPACE) -o jsonpath='{.data.token}' -n $NAMESPACE | base64 --decode )

2. get a list of PXC clusters

   .. code-block:: bash

>       # run "kubectl --v=9 get pxc" command find curl command in the output

Reviewer: Kubectl Command:

```suggestion
      kubectl --v=9 get pxc
```

I think we can remove `--v=9` from the commands. It was just an example for developers how to get 'curl' ... @delgod  what do you think?

Reviewer 2: have sense

PXC API Documentation
=====================

.. raw:: html

   <style>
   
   .toggle {
        background: none repeat scroll 0 0 #ffebcc;
        padding: 12px;
        line-height: 24px;
        margin-bottom: 24px;
    }
   
   .toggle .header {
       display: block;
       clear: both;
       cursor: pointer;
   }
   
   .toggle .header:after {
       content: " ▶";
   }
   
   .toggle .header.open:after {
       content: " ▼";
   }
   </style>

Percona Operator for Percona XtraDB Cluster provides an `aggregation-layer extension for the Kubernetes API <https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/>`_. Please refer to the
`official Kubernetes API documentation <https://kubernetes.io/docs/reference/>`_ on the API access and usage details.
The following subsections describe the Percona XtraDB Cluster API provided by the Operator.

.. contents:: :local:

Prerequisites
-------------

1. prepare

   .. code-block:: yaml

      # set correct API adress
      KUBE_CLUSTER=$(kubectl config view --minify -o jsonpath='{.clusters[0].name}')

>       API_SERVER=$(kubectl config view -o jsonpath="{.clusters[?(@.name==\"$KUBE_CLUSTER\")].cluster.server}")

Reviewer: this var contains `https://IP` so we need to remove `https://` e.g. using the following `API_SERVER=$(kubectl config view -o jsonpath="{.clusters[?(@.name==\"$KUBE_CLUSTER\")].cluster.server}" | tr -d 'https://')` and fix line https://github.com/percona/percona-xtradb-cluster-operator/pull/518/files#diff-cc08f3ae0f8900da083eec2e9c304d71R56 or we need to fix all document by removing `https:// `  from e.g. https://github.com/percona/percona-xtradb-cluster-operator/pull/518/files#diff-cc08f3ae0f8900da083eec2e9c304d71R77

Reviewer: As I can see It was fixed for psmdb but not for pxc  doc :)

> 

Reviewer: As I can see for now we work with default namespace but we need to show how to work with different namespaces . So I suggest adding the following:
```
NAMESPACE='pxc'
kubectl create namespace $NAMESPACE
```

      # create service account and get token

>       kubectl apply -f deploy/crd.yaml -f deploy/rbac.yaml

Reviewer: ```suggestion
      kubectl apply -f deploy/crd.yaml -f deploy/rbac.yaml -n $NAMESPACE
```

>       KUBE_TOKEN=$(kubectl get secret $(kubectl get serviceaccount percona-xtradb-cluster-operator -o jsonpath='{.secrets[0].name}') -o jsonpath='{.data.token}' | base64 --decode )

Reviewer: ```suggestion
      KUBE_TOKEN=$(kubectl get secret $(kubectl get serviceaccount percona-xtradb-cluster-operator -o jsonpath='{.secrets[0].name}' -n $NAMESPACE) -o jsonpath='{.data.token}' -n $NAMESPACE | base64 --decode )
```

2. get a list of PXC clusters

   .. code-block:: bash

      # run "kubectl --v=9 get pxc" command find curl command in the output

>       curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "$APISERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool

Reviewer: ```suggestion
      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
```

       content: " ▶";
   }
   
   .toggle .header.open:after {
       content: " ▼";
   }
   </style>

Percona Operator for Percona XtraDB Cluster provides an `aggregation-layer extension for the Kubernetes API <https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/>`_. Please refer to the
`official Kubernetes API documentation <https://kubernetes.io/docs/reference/>`_ on the API access and usage details.
The following subsections describe the Percona XtraDB Cluster API provided by the Operator.

.. contents:: :local:

Prerequisites
-------------

1. Create the namespace name you will use, if not exist:

   .. code-block:: yaml

      kubectl create namespace my-namespace-name

   Trying to create an already-existing namespace will show you a
   self-explanatory error message. Also, you can use the ``defalut`` namespace.

   .. note:: In this document ``default`` namespace is used in all examples.
      Substitute ``default`` with your namespace name if you use a different
      one.

2. Prepare

   .. code-block:: yaml

      # set correct API address
      KUBE_CLUSTER=$(kubectl config view --minify -o jsonpath='{.clusters[0].name}')

>       API_SERVER=$(kubectl config view -o jsonpath="{.clusters[?(@.name==\"$KUBE_CLUSTER\")].cluster.server}")

Reviewer: ```suggestion
      API_SERVER=$(kubectl config view -o jsonpath="{.clusters[?(@.name==\"$KUBE_CLUSTER\")].cluster.server}" | tr -d 'https://')
```

      # create service account and get token
      kubectl apply -f deploy/crd.yaml -f deploy/rbac.yaml
      KUBE_TOKEN=$(kubectl get secret $(kubectl get serviceaccount percona-xtradb-cluster-operator -o jsonpath='{.secrets[0].name}' -n default) -o jsonpath='{.data.token}' -n default | base64 --decode )
      

3. Get a list of PXC clusters

   .. code-block:: bash

      kubectl --v=9 get pxc

>       curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool

Reviewer: ```suggestion
      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
```

> 3. Get a list of PXC clusters
> 
>    .. code-block:: bash
> 
>       kubectl get pxc
>       curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool

Reviewer 2: we should remove these lines or rework in the following way
```suggestion
3. Check if we have correct values in API_SERVER and KUBE_TOKEN variables

   .. code-block:: bash

      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
      
      expected output is
    
     .. code-block:: bash

    {{...@fiowro, please PUT REAL OUTPUT HERE...}}
```

> 3. Get a list of PXC clusters
> 
>    .. code-block:: bash
> 
>       kubectl --v=9 get pxc
>       curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool

Reviewer 2: ```suggestion
3. Check if we have correct values in API_SERVER and KUBE_TOKEN variabled

   .. code-block:: bash

      curl -k -XGET  -H "Authorization: Bearer $KUBE_TOKEN" "https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters?limit=500" | python -mjson.tool
```

**Authentication:**

.. code-block:: bash

   Authorization: Bearer $KUBE_TOKEN

**cURL Request:**

.. code-block:: bash

   curl -k -v -XPATCH 'https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters/cluster1' \
               -H "Authorization: Bearer $KUBE_TOKEN" \
               -H "Content-Type: application/merge-patch+json" 
               -H "Accept: application/json" \
               -d '{  
                     "spec": {"pxc":{ "size": "5" }  
                     }}'

**Request Body:**

.. container:: toggle

   .. container:: header

      JSON:

   .. include:: ./assets/code/api-scale-cluster-request-json.txt

**Input:**

   **spec**:

   pxc

   1. size (Int or String, Defaults: 3): ``Specifiy the sie of the pxc cluster to scale up or down to``

**Response:**

.. container:: toggle

   .. container:: header

      JSON:

   .. include:: ./assets/code/api-scale-cluster-response-json.txt

Update PXC cluster image
------------------------

**Description:**

.. code-block:: bash

   Change the image of pxc containers inside the cluster

**Kubectl Command:**

.. code-block:: bash

   kubectl --v=9 patch pxc cluster1 --type=merge --patch '{  

>    "spec": {"pxc":{ "image": "percona/percona-xtradb-cluster-operator:1.5.0-pxc5.7" }  

Reviewer: as you know we do not have  such image starting from 1.5.0.  Please use `percona/percona-xtradb-cluster:5.7.30-31.43`

   }}'

**URL:**

.. code-block:: bash

   https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters/cluster1

**Authentication:**

.. code-block:: bash

   Authorization: Bearer $KUBE_TOKEN

**cURL Request:**

.. code-block:: bash

   curl -k -v -XPATCH 'https://$API_SERVER/apis/pxc.percona.com/v1/namespaces/default/perconaxtradbclusters/cluster1' \
               -H "Authorization: Bearer $KUBE_TOKEN" \
               -H "Accept: application/json" \
               -H "Content-Type: application/merge-patch+json" 
               -d '{  

>                  "spec": {"pxc":{ "image": "percona/percona-xtradb-cluster-operator:1.5.0-pxc5.7" }

Reviewer: ```suggestion
                 "spec": {"pxc":{ "image": "percona/percona-xtradb-cluster:5.7.30-31.43" }
```
